libcaf_core/caf/intrusive/linked_list.hpp

```diff
@@ -0,0 +1,283 @@
+
+  /// Returns a pointer to the first element.
+  pointer front() noexcept {
+    return promote(head_.next);
+  }
+
+  /// Returns a pointer to the last element.
+  pointer back() noexcept {
+    CAF_ASSERT(head_.next != &tail_);
+    return promote(tail_.next);
+  }
+
+  /// Like `front`, but returns `nullptr` if the list is empty.
+  pointer peek() noexcept {
+    return size_ > 0 ? front() : nullptr;
+  }
+
+  // -- insertion --------------------------------------------------------------
+
+  /// Appends `ptr` to the list.
+  /// @pre `ptr != nullptr`
+  void push_back(pointer ptr) noexcept {
+    CAF_ASSERT(ptr != nullptr);
+    tail_.next->next = ptr;
+    tail_.next = ptr;
+    ptr->next = &tail_;
+    ++size_;
+  }
+
+  /// Appends `ptr` to the list.
+  /// @pre `ptr != nullptr`
+  void push_back(unique_pointer ptr) noexcept {
+    push_back(ptr.release());
+  }
+
+  /// Creates a new element from `xs...` and calls `push_back` with it.
+  template <class... Ts>
+  void emplace_back(Ts&&... xs) {
+    push_back(new value_type(std::forward<Ts>(xs)...));
+  }
+
+  void push_front(pointer ptr) noexcept {
+    CAF_ASSERT(ptr != nullptr);
+    if (empty()) {
+      push_back(ptr);
+      return;
+    }
+    ptr->next = head_.next;
+    head_.next = ptr;
+    ++size_;
+  }
+
+  void push_front(unique_pointer ptr) noexcept {
+    push_front(ptr.release());
+  }
+
+  /// Creates a new element from `xs...` and calls `push_front` with it.
+  template <class... Ts>
+  void emplace_front(Ts&&... xs) {
+    push_front(new value_type(std::forward<Ts>(xs)...));
+  }
+
+  /// Inserts `ptr` after `pos`.
+  iterator insert_after(iterator pos, pointer ptr) {
+    CAF_ASSERT(pos != end());
+    CAF_ASSERT(ptr != nullptr);
+    auto next = pos->next;
+    ptr->next = next;
+    pos->next = ptr;
+    if (next == &tail_)
+      tail_.next = ptr;
+    ++size_;
+    return iterator{ptr};
+  }
+
+  // -- algorithms -------------------------------------------------------------
+
+  /// Tries to find an element in the list that matches the given predicate.
+  template <class Predicate>
+  pointer find_if(Predicate pred) {
+    for (auto i = begin(); i != end(); ++i)
+      if (pred(*i))
+        return promote(i.ptr);
+    return nullptr;
+  }
+
+  /// Transfers ownership of all elements to `fn`.
+  template <class F>
+  void drain(F fn) {
+    for (auto i = head_.next; i != &tail_;) {
+      auto ptr = i;
+      i = i->next;
+      fn(promote(ptr));
+    }
+    init();
+  }
+
+private:
+  /// Restores a consistent, empty state.
+  void init() noexcept {
+    head_.next = &tail_;
+    tail_.next = &head_;
+    size_ = 0;
+  }
+
+  // -- member variables ------------------------------------------------------
+
+  /// Dummy before-the-first-element node.
+  node_type head_;
+
+  /// Dummy past-the-last-element node. The `tail_->next` pointer is pointing to
+  /// the last element in the list or to `head_` if the list is empty.
+  node_type tail_;
```

Comment:
Could we add these changes, and then skip calling `init` in the default and rvalue ctors?

```suggestion
  node_type head_ = &tail_;

  /// Dummy past-the-last-element node. The `tail_->next` pointer is pointing to
  /// the last element in the list or to `head_` if the list is empty.
  node_type tail_ = &head_;
```